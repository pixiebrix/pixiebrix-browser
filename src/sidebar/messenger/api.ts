/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/* Do not use `registerMethod` in this file */
import { getMethod, getNotifier } from "webext-messenger";

const target = { tabId: "this", page: "/sidebar.html" } as const;
const sidebarInThisTab = {
  renderPanels: getMethod("SIDEBAR_RENDER_PANELS", target),
  activatePanel: getMethod("SIDEBAR_ACTIVATE_PANEL", target),
  showForm: getMethod("SIDEBAR_SHOW_FORM", target),
  hideForm: getMethod("SIDEBAR_HIDE_FORM", target),
  pingSidebar: getMethod("SIDEBAR_PING", target),
  showTemporaryPanel: getMethod("SIDEBAR_SHOW_TEMPORARY_PANEL", target),
  updateTemporaryPanel: getNotifier("SIDEBAR_UPDATE_TEMPORARY_PANEL", target),
  hideTemporaryPanel: getMethod("SIDEBAR_HIDE_TEMPORARY_PANEL", target),
  showModActivationPanel: getMethod("SIDEBAR_SHOW_ACTIVATE_RECIPE", target),
  hideModActivationPanel: getMethod("SIDEBAR_HIDE_ACTIVATE_RECIPE", target),
};

export default sidebarInThisTab;
