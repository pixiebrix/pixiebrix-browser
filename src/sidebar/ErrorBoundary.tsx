/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { Component } from "react";
import { isExtensionContext } from "webext-detect-page";
import { getErrorMessage } from "@/errors/errorHelpers";
import reportError from "@/telemetry/reportError";
import { UnknownObject } from "@/types";
import { isEmpty } from "lodash";
import { faRedo } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { Alert, Button } from "react-bootstrap";
import { whoAmI } from "@/background/messenger/api";
import { reloadSidebar } from "@/contentScript/messenger/api";

interface State {
  hasError: boolean;
  errorMessage: string;
  stack: string;
}

class ErrorBoundary extends Component<UnknownObject, State> {
  constructor(props: UnknownObject) {
    super(props);
    this.state = { hasError: false, errorMessage: undefined, stack: undefined };
  }

  static getDerivedStateFromError(error: Error) {
    // Update state so the next render will show the fallback UI.
    return {
      hasError: true,
      errorMessage: getErrorMessage(error),
      stack: error.stack,
    };
  }

  override componentDidCatch(error: Error): void {
    if (isExtensionContext()) {
      reportError(error);
    }
  }

  async reloadSidebar() {
    const sidebar = await whoAmI();
    await reloadSidebar({ tabId: sidebar.tab.id });
  }

  override render(): React.ReactNode {
    if (this.state.hasError) {
      return (
        <div className="p-3">
          <Alert variant="danger">
            <Alert.Heading>Something went wrong</Alert.Heading>
            {!isEmpty(this.state.errorMessage) && (
              <>
                <p>{this.state.errorMessage}</p>
              </>
            )}

            <p>Please close and re-open the sidebar panel.</p>

            <div>
              <Button variant="light" onClick={this.reloadSidebar}>
                <FontAwesomeIcon icon={faRedo} /> Reload Sidebar
              </Button>
            </div>
          </Alert>

          {this.state.stack && (
            <pre className="mt-2 small text-secondary">
              {this.state.stack
                // In the app
                .replaceAll(location.origin + "/", "")
                // In the content script
                .replaceAll(
                  `chrome-extension://${process.env.CHROME_EXTENSION_ID}/`,
                  ""
                )}
            </pre>
          )}
        </div>
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
