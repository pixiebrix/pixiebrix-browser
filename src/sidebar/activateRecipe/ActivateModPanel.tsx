/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, {
  useCallback,
  useEffect,
  useMemo,
  useReducer,
  useRef,
} from "react";
import { type RegistryId } from "@/types/registryTypes";
import Loader from "@/components/Loader";
import activationCompleteImage from "@img/blueprint-activation-complete.png";
import styles from "./ActivateModPanel.module.scss";
import AsyncButton from "@/components/AsyncButton";
import { useDispatch } from "react-redux";
import sidebarSlice from "@/sidebar/sidebarSlice";
import { reloadMarketplaceEnhancements as reloadMarketplaceEnhancementsInContentScript } from "@/contentScript/messenger/api";
import { getTopLevelFrame } from "webext-messenger";
import cx from "classnames";
import { isEmpty } from "lodash";
import ActivateModInputs from "@/sidebar/activateRecipe/ActivateModInputs";
import useQuickbarShortcut from "@/hooks/useQuickbarShortcut";
import { Button } from "react-bootstrap";
import useActivateRecipe from "@/activation/useActivateRecipe";
import { type WizardValues } from "@/activation/wizardTypes";
import { createSlice, type PayloadAction } from "@reduxjs/toolkit";
import useActivateRecipeWizard, {
  type UseActivateRecipeWizardResult,
} from "@/activation/useActivateRecipeWizard";
import RequireMods, {
  type RequiredModDefinition,
} from "@/sidebar/activateRecipe/RequireMods";
import { persistor } from "@/sidebar/store";
import { checkModDefinitionPermissions } from "@/modDefinitions/modDefinitionPermissionsHelpers";
import AsyncStateGate from "@/components/AsyncStateGate";
import { type ModDefinition } from "@/types/modDefinitionTypes";

import { openShortcutsTab, SHORTCUTS_URL } from "@/utils/extensionUtils";

const { actions } = sidebarSlice;

type ActivationState = {
  /**
   * True if recipe needs permissions to activate, false if it does not, null if not calculated yet.
   */
  needsPermissions: boolean | null;
  /**
   * True if the recipe is manually or automatically activating.
   */
  isActivating: boolean;
  /**
   * True if the recipe has been activated.
   */
  isActivated: boolean;
  /**
   * The error message if the recipe failed to activate, or null if there is no error
   */
  activationError: string | null;
};

const initialState: ActivationState = {
  isActivating: false,
  isActivated: false,
  // Permissions have not been checked yet
  needsPermissions: null,
  activationError: null,
};

const activationSlice = createSlice({
  name: "activationSlice",
  initialState,
  reducers: {
    setNeedsPermissions(state, action: PayloadAction<boolean>) {
      state.needsPermissions = action.payload;
    },
    activateStart(state) {
      state.isActivating = true;
      state.activationError = null;
    },
    activateSuccess(state) {
      state.isActivating = false;
      state.isActivated = true;
    },
    activateError(state, action: PayloadAction<string>) {
      state.activationError = action.payload;
      state.isActivating = false;
    },
  },
});

const { setNeedsPermissions, activateStart, activateSuccess, activateError } =
  activationSlice.actions;

async function reloadMarketplaceEnhancements() {
  const topFrame = await getTopLevelFrame();
  // Make sure the content script has the most recent state of the store before reloading.
  // Prevents race condition where the content script reloads before the store is persisted.
  await persistor.flush();
  void reloadMarketplaceEnhancementsInContentScript(topFrame);
}

const ShortcutKeys: React.FC<{ shortcut: string | null }> = ({ shortcut }) => {
  const separator = shortcut?.includes("+") ? "+" : "";
  const shortcutKeys = shortcut?.split(separator) ?? [];
  return (
    <div className={styles.shortcutContainer}>
      {shortcutKeys.map((key, index) => (
        <React.Fragment key={key}>
          {index > 0 && <span>&nbsp;&nbsp;+&nbsp;&nbsp;</span>}
          <span className={styles.shortcutKey}>{key.trim()}</span>
        </React.Fragment>
      ))}
    </div>
  );
};

const ModName: React.FC<{ modDefinition: ModDefinition }> = ({
  modDefinition,
}) => <div className={styles.modName}>{modDefinition.metadata.name}</div>;

export const SuccessMessage: React.FC<{
  includesQuickBar: boolean;
  numMods: number;
}> = ({ numMods, includesQuickBar }) => {
  const { shortcut } = useQuickbarShortcut();

  if (includesQuickBar && isEmpty(shortcut)) {
    return (
      <span>
        Now just{" "}
        <Button
          variant="link"
          href={SHORTCUTS_URL}
          onClick={(event) => {
            // `react-bootstrap` will render as an anchor tag when href is set
            // Can't link to chrome:// URLs directly
            event.preventDefault();
            void openShortcutsTab();
          }}
          className={styles.configureLink}
        >
          Configure your Quick Bar shortcut
        </Button>{" "}
        to access this mod.
      </span>
    );
  }

  if (includesQuickBar) {
    return (
      <>
        <div>
          Launch {numMods > 1 ? "them" : "it"} using your Quick Bar shortcut
        </div>
        <ShortcutKeys shortcut={shortcut} />
        <Button
          variant="link"
          href={SHORTCUTS_URL}
          onClick={(event) => {
            // `react-bootstrap` will render as an anchor tag when href is set
            // Can't link to chrome:// URLs directly
            event.preventDefault();
            void openShortcutsTab();
          }}
        >
          Change the Quick Bar shortcut.
        </Button>
      </>
    );
  }

  return <div>Go try it out now, or activate another mod.</div>;
};

export const SuccessPanel: React.FC<{
  title: React.ReactNode;
  includesQuickBar: boolean;
  numMods: number;
  handleActivationDecision: () => Promise<void>;
}> = ({ title, includesQuickBar, handleActivationDecision, numMods }) => (
  <div className={styles.root}>
    <div className={cx("scrollable-area", styles.content)}>
      <h1>Well done!</h1>
      <img src={activationCompleteImage} alt="" width={300} />
      <div className={styles.textContainer}>
        {title}
        <br />
        <SuccessMessage includesQuickBar={includesQuickBar} numMods={numMods} />
      </div>
    </div>
    <div className={styles.footer}>
      <AsyncButton onClick={handleActivationDecision}>Ok</AsyncButton>
    </div>
  </div>
);

const ActivateRecipePanelContent: React.FC<
  RequiredModDefinition & UseActivateRecipeWizardResult
> = ({
  modDefinition,
  includesQuickBar,
  requiresConfiguration,
  wizardSteps,
  initialValues,
  validationSchema,
  isActive,
}) => {
  const reduxDispatch = useDispatch();
  const marketplaceActivateRecipe = useActivateRecipe("marketplace");

  const [state, stateDispatch] = useReducer(
    activationSlice.reducer,
    initialState
  );

  async function handleActivationDecision() {
    reduxDispatch(actions.hideModActivationPanel());
  }

  const formValuesRef = useRef<WizardValues>(initialValues);

  async function checkPermissions() {
    const configuredDependencies =
      formValuesRef.current.integrationDependencies.filter(
        ({ config }) => !isEmpty(config)
      );

    const { hasPermissions } = await checkModDefinitionPermissions(
      modDefinition,
      configuredDependencies
    );

    stateDispatch(setNeedsPermissions(!hasPermissions));
  }

  const onChange = (values: WizardValues) => {
    formValuesRef.current = values;
    void checkPermissions();
  };

  const activateRecipe = useCallback(async () => {
    if (state.isActivating || state.isActivated) {
      // Prevent double-activation
      return;
    }

    stateDispatch(activateStart());

    const { success, error } = await marketplaceActivateRecipe(
      formValuesRef.current,
      modDefinition
    );

    if (success) {
      stateDispatch(activateSuccess());
      void reloadMarketplaceEnhancements();
    } else {
      stateDispatch(activateError(error));
    }
  }, [
    marketplaceActivateRecipe,
    modDefinition,
    state.isActivated,
    state.isActivating,
  ]);

  // Check permissions on mount, to determine if the recipe can be auto-activated, and whether to show the
  // permissions request information UI.
  useEffect(() => {
    void checkPermissions();
    // eslint-disable-next-line react-hooks/exhaustive-deps -- only run on mount
  }, []);

  // Trigger auto-activation if the recipe does not require permissions or user configuration, and there's no error.
  // Additionally, always show the options if this is a reactivation.  (Reactivation, in practice, is used to
  // reconfigure mod options)
  useEffect(() => {
    if (
      state.needsPermissions != null &&
      !state.needsPermissions &&
      !requiresConfiguration &&
      !state.activationError &&
      !isActive
    ) {
      // State is checked inside activateRecipe to prevent double-activation
      void activateRecipe();
    }
  }, [
    initialValues,
    activateRecipe,
    requiresConfiguration,
    state.needsPermissions,
    state.activationError,
    isActive,
  ]);

  // Show loader if panel is determining if it can auto-activate, or if it's activating.
  if (state.needsPermissions == null || state.isActivating) {
    return <Loader />;
  }

  if (state.isActivated) {
    return (
      <SuccessPanel
        title={
          <>
            <ModName modDefinition={modDefinition} />
            <div>is ready to use!</div>
          </>
        }
        numMods={1}
        includesQuickBar={includesQuickBar}
        handleActivationDecision={handleActivationDecision}
      />
    );
  }

  return (
    <div className={styles.root}>
      <ActivateModInputs
        recipe={modDefinition}
        optionsWizardStep={wizardSteps.find(({ key }) => key === "options")}
        initialValues={initialValues}
        onChange={onChange}
        validationSchema={validationSchema}
        onClickCancel={handleActivationDecision}
        needsPermissions={state.needsPermissions}
        header={
          <>
            <ModName modDefinition={modDefinition} />
            <p>
              {
                "We're almost there. This mod has a few settings to configure before using. You can always change these later."
              }
            </p>
          </>
        }
        formValuesRef={formValuesRef}
        onClickSubmit={() => {
          void activateRecipe();
        }}
        activationError={state.activationError}
      />
    </div>
  );
};

const ActivateModWizardPanel: React.FC<RequiredModDefinition> = (modState) => {
  const wizardState = useActivateRecipeWizard(
    modState.modDefinition,
    modState.defaultAuthOptions
  );
  return (
    <AsyncStateGate state={wizardState}>
      {({ data: wizardResult }) => (
        <ActivateRecipePanelContent {...modState} {...wizardResult} />
      )}
    </AsyncStateGate>
  );
};

/**
 * React Component Panel for activating a single mod, which may require end-user configuration.
 * @param modId the mod id
 */
const ActivateModPanel: React.FC<{ modId: RegistryId }> = ({ modId }) => {
  // Memoize to array reference doesn't change on re-render
  const modIds = useMemo(() => [modId], [modId]);

  return (
    <RequireMods modIds={modIds}>
      {(modDefinitions) => <ActivateModWizardPanel {...modDefinitions[0]} />}
    </RequireMods>
  );
};

export default ActivateModPanel;
