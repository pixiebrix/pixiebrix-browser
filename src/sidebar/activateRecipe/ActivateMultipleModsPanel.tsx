/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useMemo } from "react";
import { type RegistryId } from "@/types/registryTypes";
import RequireMods, {
  type RequiredModDefinition,
} from "@/sidebar/activateRecipe/RequireMods";
import AsyncStateGate from "@/components/AsyncStateGate";
import { getOptionsValidationSchema } from "@/hooks/useAsyncRecipeOptionsValidationSchema";
import useDatabaseOptions from "@/hooks/useDatabaseOptions";
import { useDispatch, useSelector } from "react-redux";
import { selectExtensions } from "@/store/extensionsSelectors";
import useDeriveAsyncState from "@/hooks/useDeriveAsyncState";
import { type Option } from "@/components/form/widgets/SelectWidget";
import { wizardStateFactory } from "@/activation/useActivateRecipeWizard";
import useActivateRecipe, {
  type ActivateResult,
} from "@/activation/useActivateRecipe";
import { SuccessPanel } from "@/sidebar/activateRecipe/ActivateModPanel";
import { getTopLevelFrame } from "webext-messenger";
import { hideSidebar } from "@/contentScript/messenger/api";
import sidebarSlice from "@/sidebar/sidebarSlice";
import { selectSidebarHasModPanels } from "@/sidebar/sidebarSelectors";

type ModResultPair = {
  mod: RequiredModDefinition;
  result: ActivateResult;
};

const MultipleSuccessPanel: React.FC<{ results: ModResultPair[] }> = ({
  results,
}) => {
  const reduxDispatch = useDispatch();
  const sidebarHasModPanels = useSelector(selectSidebarHasModPanels);

  async function handleActivationDecision() {
    reduxDispatch(sidebarSlice.actions.hideModActivationPanel());

    if (!sidebarHasModPanels) {
      const topFrame = await getTopLevelFrame();
      void hideSidebar(topFrame);
    }
  }

  return (
    <SuccessPanel
      title="Your mods are ready to use!"
      numMods={results.length}
      includesQuickBar={results.some((x) => x.mod.includesQuickBar)}
      handleActivationDecision={handleActivationDecision}
    />
  );
};

/**
 * React Component Panel to automatically activate multiple mods and show a success message.
 * @param mods mod definitions supporting automatic activation
 */
const AutoActivatePanel: React.FC<{ mods: RequiredModDefinition[] }> = ({
  mods,
}) => {
  // Assume mods work without all permissions. Currently, the only optional permission is `clipboardWrite`, which isn't
  // actually enforced by Chrome. (Mods can still copy to the clipboard.). The only way a mod would not have all
  // permissions is if their Enterprise policy has disabled some permissions.
  const activate = useActivateRecipe("marketplace", {
    checkPermissions: false,
  });

  // Only activate new mods that the user does not already have activated. If there are updates available, the
  // user will be prompted to update according to marketplace mod updater rules.
  const newMods = useMemo(() => mods.filter((x) => !x.isActive), [mods]);
  const activatedModComponents = useSelector(selectExtensions);
  const databaseOptionsState = useDatabaseOptions({ refetchOnMount: true });

  // A bit hacky -- using useDeriveAsyncState to automatically activate the mods on mount
  const activationResultState = useDeriveAsyncState(
    databaseOptionsState,
    async (databaseOptions: Option[]) => {
      if (newMods.some((x) => x.requiresConfiguration)) {
        throw new Error(
          "One or more mods require configuration. Activate the mods individually to configure them."
        );
      }

      return Promise.all(
        newMods.map(async (mod) => {
          const optionsValidationSchema = await getOptionsValidationSchema(
            mod.modDefinition.options?.schema
          );

          const wizard = wizardStateFactory({
            modDefinition: mod.modDefinition,
            defaultAuthOptions: mod.defaultAuthOptions,
            databaseOptions,
            optionsValidationSchema,
            installedExtensions: activatedModComponents,
          });

          const result = await activate(
            wizard.initialValues,
            mod.modDefinition
          );

          if (result.error) {
            throw new Error(
              `Error activating ${mod.modDefinition.metadata.name}`,
              { cause: new Error(result.error) }
            );
          }

          return {
            result,
            mod,
          };
        })
      );
    }
  );

  return (
    <AsyncStateGate state={activationResultState}>
      {({ data: results }) => <MultipleSuccessPanel results={results} />}
    </AsyncStateGate>
  );
};

/**
 * React Component Panel to automatically activate multiple mods and show a success message.
 * @param modId the mod id
 *
 * @since 1.7.35
 */
const ActivateMultipleModsPanel: React.FC<{ modIds: RegistryId[] }> = ({
  modIds,
}) => (
  <RequireMods modIds={modIds}>
    {(mods) => <AutoActivatePanel mods={mods} />}
  </RequireMods>
);

export default ActivateMultipleModsPanel;
