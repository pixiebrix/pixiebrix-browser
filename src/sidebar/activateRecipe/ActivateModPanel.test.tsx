/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { useRequiredModDefinitions } from "@/modDefinitions/modDefinitionHooks";
import { render, screen } from "@/sidebar/testHelpers";
import ActivateModPanel from "@/sidebar/activateRecipe/ActivateModPanel";
import sidebarSlice from "@/sidebar/sidebarSlice";
import { waitForEffect } from "@/testUtils/testHelpers";
import { propertiesToSchema } from "@/validators/generic";
import registerDefaultWidgets from "@/components/fields/schemaFields/widgets/registerDefaultWidgets";
import useQuickbarShortcut from "@/hooks/useQuickbarShortcut";
import { type ModDefinition } from "@/types/modDefinitionTypes";
import { valueToAsyncCacheState } from "@/utils/asyncStateUtils";
import { validateRegistryId } from "@/types/helpers";
import { checkModDefinitionPermissions } from "@/modDefinitions/modDefinitionPermissionsHelpers";
import { appApiMock, onDeferredGet } from "@/testUtils/appApiMock";
import {
  getModDefinitionWithBuiltInServiceAuths,
  defaultModDefinitionFactory,
} from "@/testUtils/factories/modDefinitionFactories";
import { sidebarEntryFactory } from "@/testUtils/factories/sidebarEntryFactories";
import {
  marketplaceListingFactory,
  modDefinitionToMarketplacePackage,
} from "@/testUtils/factories/marketplaceFactories";
import * as messengerApi from "@/contentScript/messenger/api";
import { selectSidebarHasModPanels } from "@/sidebar/sidebarSelectors";
import userEvent from "@testing-library/user-event";
import ActivateMultipleModsPanel from "@/sidebar/activateRecipe/ActivateMultipleModsPanel";
import ErrorBoundary from "@/sidebar/ErrorBoundary";
import { includesQuickBarStarterBrick } from "@/utils/modDefinitionUtils";

jest.mock("@/modDefinitions/modDefinitionHooks", () => ({
  useRequiredModDefinitions: jest.fn(),
}));

jest.mock("@/sidebar/sidebarSelectors", () => ({
  selectSidebarHasModPanels: jest.fn(),
}));

const useRequiredModDefinitionsMock = jest.mocked(useRequiredModDefinitions);
const checkModDefinitionPermissionsMock = jest.mocked(
  checkModDefinitionPermissions
);
const selectSidebarHasModPanelsMock = jest.mocked(selectSidebarHasModPanels);
const hideSidebarSpy = jest.spyOn(messengerApi, "hideSidebar");

jest.mock("@/utils/modDefinitionUtils", () => {
  const actualUtils = jest.requireActual("@/utils/modDefinitionUtils");

  return {
    __esModule: true,
    ...actualUtils,
    includesQuickBarStarterBrick: jest.fn().mockResolvedValue(true),
  };
});

const includesQuickBarMock = jest.mocked(includesQuickBarStarterBrick);

jest.mock("@/registry/internal", () => ({
  // We're also mocking all the functions that this output is passed to, so we can return empty array
  resolveRecipe: jest.fn().mockResolvedValue([]),
}));

jest.mock("@/hooks/useQuickbarShortcut", () => ({
  __esModule: true,
  default: jest.fn(),
}));

const useQuickbarShortcutMock = jest.mocked(useQuickbarShortcut);

jest.mock("@/activation/useActivateRecipe", () => ({
  __esModule: true,
  default: jest.fn().mockReturnValue(async () => ({ success: true })),
}));

beforeAll(() => {
  registerDefaultWidgets();
});

function setupMocksAndRender(
  modDefinitionOverride?: Partial<ModDefinition>,
  { componentOverride }: { componentOverride?: React.ReactElement } = {}
) {
  const modDefinition = defaultModDefinitionFactory({
    ...modDefinitionOverride,
    metadata: {
      id: validateRegistryId("test-mod"),
      name: "Test Mod",
    },
  });
  useRequiredModDefinitionsMock.mockReturnValue(
    valueToAsyncCacheState([modDefinition])
  );
  const listing = marketplaceListingFactory({
    // Consistent user-visible name for snapshots
    package: modDefinitionToMarketplacePackage(modDefinition),
  });

  // Tests can override by calling before setupMocksAndRender
  appApiMock.onGet("/api/marketplace/listings/").reply(200, [listing]);
  appApiMock.onGet().reply(200, []);

  const entry = sidebarEntryFactory("activateMods", {
    modIds: [modDefinition.metadata.id],
    heading: "Activate Mod",
  });

  const element = componentOverride ?? (
    <ActivateModPanel modId={modDefinition.metadata.id} />
  );

  return render(element, {
    setupRedux(dispatch) {
      dispatch(sidebarSlice.actions.showModActivationPanel(entry));
    },
  });
}

beforeEach(() => {
  appApiMock.reset();
  selectSidebarHasModPanelsMock.mockReset();
  hideSidebarSpy.mockReset();

  includesQuickBarMock.mockResolvedValue(false);

  useQuickbarShortcutMock.mockReturnValue({
    shortcut: null,
    isConfigured: false,
  });

  checkModDefinitionPermissionsMock.mockResolvedValue({
    hasPermissions: true,
    permissions: {},
  });
});

describe("ActivateRecipePanel", () => {
  it("renders with options, permissions info", async () => {
    jest.mocked(checkModDefinitionPermissions).mockResolvedValue({
      hasPermissions: false,
      permissions: { origins: ["https://newurl.com"] },
    });

    const rendered = setupMocksAndRender({
      options: {
        schema: propertiesToSchema({
          foo: {
            type: "string",
          },
          bar: {
            type: "number",
          },
          testDatabase: {
            $ref: "https://app.pixiebrix.com/schemas/database#",
            title: "Test Database",
          },
        }),
      },
    });

    await waitForEffect();

    rendered.debug();

    expect(rendered.asFragment()).toMatchSnapshot();
  });

  it("activates basic recipe automatically and renders well-done page", async () => {
    const rendered = setupMocksAndRender();

    await waitForEffect();

    expect(rendered.asFragment()).toMatchSnapshot();
  });

  it("activates basic recipe with empty options structure automatically and renders well-done page", async () => {
    const rendered = setupMocksAndRender({
      options: {
        schema: {},
      },
    });

    await waitForEffect();

    expect(rendered.asFragment()).toMatchSnapshot();
  });

  it("activates recipe with database preview automatically and renders well-done page", async () => {
    const rendered = setupMocksAndRender({
      options: {
        schema: propertiesToSchema({
          testDatabase: {
            $ref: "https://app.pixiebrix.com/schemas/database#",
            title: "Database",
            format: "preview",
          },
        }),
      },
    });

    await waitForEffect();

    expect(rendered.asFragment()).toMatchSnapshot();
  });

  it("renders well-done page for quick bar mod shortcut not configured", async () => {
    includesQuickBarMock.mockResolvedValue(true);

    const rendered = setupMocksAndRender();

    await waitForEffect();

    expect(rendered.asFragment()).toMatchSnapshot();
  });

  it("renders well-done page for quick bar mod shortcut is configured on MacOS", async () => {
    includesQuickBarMock.mockResolvedValue(true);

    useQuickbarShortcutMock.mockReturnValue({
      shortcut: "âŒ˜M",
      isConfigured: true,
    });

    const rendered = setupMocksAndRender();

    await waitForEffect();

    expect(rendered.asFragment()).toMatchSnapshot();
  });

  it("renders well-done page for quick bar mod shortcut is configured on Windows", async () => {
    includesQuickBarMock.mockResolvedValue(true);

    useQuickbarShortcutMock.mockReturnValue({
      shortcut: "Ctrl+M",
      isConfigured: true,
    });

    const rendered = setupMocksAndRender();

    await waitForEffect();

    expect(rendered.asFragment()).toMatchSnapshot();
  });

  it("renders with service configuration if no built-in service configs available", async () => {
    const { modDefinition } = getModDefinitionWithBuiltInServiceAuths();

    const rendered = setupMocksAndRender(modDefinition);

    await waitForEffect();

    expect(rendered.asFragment()).toMatchSnapshot();
    expect(
      rendered.container.querySelector(".actionButton")
    ).not.toBeDisabled();
  });

  it("activates recipe with built-in services automatically and renders well-done page", async () => {
    const { modDefinition, builtInServiceAuths } =
      getModDefinitionWithBuiltInServiceAuths();

    appApiMock.onGet("/api/services/shared/").reply(200, builtInServiceAuths);

    const rendered = setupMocksAndRender(modDefinition);

    await waitForEffect();

    expect(rendered.asFragment()).toMatchSnapshot();
  });

  it("doesn't flicker while built-in auths are loading", async () => {
    const { modDefinition } = getModDefinitionWithBuiltInServiceAuths();

    onDeferredGet("/api/services/shared/");

    const rendered = setupMocksAndRender(modDefinition);

    await waitForEffect();

    expect(rendered.getByTestId("loader")).not.toBeNull();
  });

  it("activating a mod closes the sidebar when there are no other mod panels to show", async () => {
    selectSidebarHasModPanelsMock.mockImplementation(() => false);

    setupMocksAndRender();
    await waitForEffect();

    await userEvent.click(screen.getByRole("button", { name: /ok/i }));

    expect(hideSidebarSpy).toHaveBeenCalledTimes(1);
  });

  it("activating a mod leaves the sidebar open when there are other mod panels to show", async () => {
    selectSidebarHasModPanelsMock.mockImplementation(() => true);

    setupMocksAndRender();
    await waitForEffect();

    await userEvent.click(screen.getByRole("button", { name: /ok/i }));

    expect(hideSidebarSpy).toHaveBeenCalledTimes(0);
  });
});

describe("ActivateMultipleModsPanel", () => {
  it("automatically activates single mod", async () => {
    const { modDefinition, builtInServiceAuths } =
      getModDefinitionWithBuiltInServiceAuths();

    appApiMock.onGet("/api/services/shared/").reply(200, builtInServiceAuths);

    const rendered = setupMocksAndRender(modDefinition, {
      componentOverride: (
        <ActivateMultipleModsPanel modIds={[modDefinition.metadata.id]} />
      ),
    });

    await waitForEffect();

    expect(rendered.asFragment()).toMatchSnapshot();
  });

  it("shows error if any mod requires configuration", async () => {
    const { modDefinition } = getModDefinitionWithBuiltInServiceAuths();

    setupMocksAndRender(modDefinition, {
      componentOverride: (
        <ErrorBoundary>
          <ActivateMultipleModsPanel modIds={[modDefinition.metadata.id]} />
        </ErrorBoundary>
      ),
    });

    await waitForEffect();

    expect(
      screen.queryByText(
        "One or more mods require configuration. Activate the mods individually to configure them."
      )
    ).toBeInTheDocument();
  });
});
