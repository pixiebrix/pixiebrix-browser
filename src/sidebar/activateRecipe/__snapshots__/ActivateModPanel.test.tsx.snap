// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivateRecipePanel activates basic recipe automatically and renders well-done page 1`] = `<DocumentFragment />`;

exports[`ActivateRecipePanel activates basic recipe with empty options structure automatically and renders well-done page 1`] = `<DocumentFragment />`;

exports[`ActivateRecipePanel activates recipe with built-in services automatically and renders well-done page 1`] = `<DocumentFragment />`;

exports[`ActivateRecipePanel activates recipe with database preview automatically and renders well-done page 1`] = `<DocumentFragment />`;

exports[`ActivateRecipePanel renders well-done page for quick bar mod shortcut is configured on MacOS 1`] = `<DocumentFragment />`;

exports[`ActivateRecipePanel renders well-done page for quick bar mod shortcut is configured on Windows 1`] = `<DocumentFragment />`;

exports[`ActivateRecipePanel renders well-done page for quick bar mod shortcut not configured 1`] = `<DocumentFragment />`;

exports[`ActivateRecipePanel renders with options, permissions info 1`] = `<DocumentFragment />`;

exports[`ActivateRecipePanel renders with service configuration if no built-in service configs available 1`] = `<DocumentFragment />`;
